Migrate to Jetpack Compose

Iamlooker opened the pull request:
Migrate:
- [x] ConsoleFragment
- [x] Migrate to Flows
- [x] ControlFragment
- [x] SettingsFragment

Iamlooker committed:
Migrate to latest `build.gradle.kts` format

Move `settings.gradle` -> `settings.gradle.kts`
Move plugin repositories `settings.gradle.kts`
Remove unused import for `jar` files

Iamlooker committed:
Remove unused `jar` imports

Iamlooker commented:
Hi, I wanted to ask, is [this](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/build.gradle.kts#L23) intentional? Because this uses the debug signing key to sign release files

Iamlooker committed:
Add Compose Dependencies

mathiascode commented:
> Hi, I wanted to ask, is [this](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/build.gradle.kts#L23) intentional? Because this uses the debug signing key to sign release files

Looks like I committed it by accident after debugging an issue in the release build. It should be removed.

Iamlooker committed:
Remove improper signing key assigning

Iamlooker commented:
> Looks like I committed it by accident after debugging an issue in the release build. It should be removed.

Done

Iamlooker commented:
Btw, why are we passing StringBuilder [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)

Either way we are converting it to string as soon as we get it [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)

Iamlooker commented:
```
Manifest merger failed : uses-sdk:minSdkVersion 16 cannot be smaller than version 21 declared in library [androidx.compose.material3:material3-android:1.2.1] C:\Users\LooKeR\.gradle\caches\transforms-4\5a41c055199354d74c6acce2479e2b57\transformed\material3-release\AndroidManifest.xml as the library might be using APIs not available in 16
	Suggestion: use a compatible library with a minSdk of at most 16,
		or increase this project's minSdk version to at least 21,
		or use tools:overrideLibrary="androidx.compose.material3" to force usage (may lead to runtime failures)
```

What should we do?

Iamlooker commented:
Also check the warning on this [page](https://developer.android.com/jetpack/androidx/versions/all-channel) They are announcing deprecation for API 21 and below

mathiascode commented:
> Btw, why are we passing StringBuilder [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)
> 
> Either way we are converting it to string as soon as we get it [here](https://github.com/cuberite/android/blob/85f4eb8974e6f2210630b85d8625eb35f5d88cb4/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L191)

No good reason, we could pass a string instead.

mathiascode commented:
> ```
> Manifest merger failed : uses-sdk:minSdkVersion 16 cannot be smaller than version 21 declared in library [androidx.compose.material3:material3-android:1.2.1] C:\Users\LooKeR\.gradle\caches\transforms-4\5a41c055199354d74c6acce2479e2b57\transformed\material3-release\AndroidManifest.xml as the library might be using APIs not available in 16
> 	Suggestion: use a compatible library with a minSdk of at most 16,
> 		or increase this project's minSdk version to at least 21,
> 		or use tools:overrideLibrary="androidx.compose.material3" to force usage (may lead to runtime failures)
> ```
> 
> What should we do?

We can bump the minimum version to 21. I believe the oldest Android versions have issues using the built-in Cuberite downloader anyway due to SSL errors.

mathiascode commented:
I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.

Iamlooker commented:
> I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.

Yes, but I did read about some changes in Android 13. I will look into this.

Iamlooker committed:
Bump `minSdk` 16 -> 21

Iamlooker committed:
Add basic theme to Jetpack Compose

Iamlooker committed:
Create utility file for formatting logs

Iamlooker committed:
Migrate Console Screen to Jetpack Compose Compose

Iamlooker commented:
![Screenshot_20240402_011658_Cuberite.jpg](https://github.com/cuberite/android/assets/60404093/1d2890db-d646-4914-8c03-6284a15ce0ce)

Current State

Iamlooker commented:
I suggest migrating LiveData to StateFlow.

This is also the suggest way to create architecture just like Jetpack Compose

Iamlooker committed:
Update Padding to mimic old layout

Iamlooker committed:
Improve Formatting logic

mathiascode commented:
> I suggest migrating LiveData to StateFlow.
> 
> This is also the suggest way to create architecture just like Jetpack Compose

If StateFlows are recommended over LiveData when using Compose, we could migrate to them. We mainly need communication between our UI and (threaded) services, which was previously achieved with LocalBroadcastManager (now deprecated).

Could the migration be done in a separate PR?

Iamlooker commented:
> Could the migration be done in a separate PR?

I will open a new PR after this one

Iamlooker committed:
Migrate `ConsoleScreen` to `ViewModel`

This helps with migration to complete compose. Right now we are using Compose in View-Fragment but once we have migrated completely, viewmodels will handle the business logic

Iamlooker committed:
Add `Button` to scroll to bottom in logs

Iamlooker commented:
> Could the migration be done in a separate PR?

I have been trying migration, but the livedata API seems to be unpredictable, or maybe its just because I am a little inexperienced with it. Should i do the migration is the same PR? or should I do that PR first and do Compose Migration later?

mathiascode commented:
> or should I do that PR first and do Compose Migration later?

Is it possible to do before the Compose migration? If the Compose migration requires more changes anyway, it would probably be better to just do it in this PR.

Iamlooker commented:
Yes, I will migrate to Flows first now :)

Iamlooker committed:
Migrate `InstallService::endedLivedata` to `StateFlow`

This performs exactly the same as the livedata implementation. Although we should not be using stateflow or even livedata to pass a one-time event. They should be used to pass the state of the application. But this seems to be not possible in the current design.

Iamlooker commented:
Btw what is [this](https://github.com/Iamlooker/cuberite-android/blob/master/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L157) tick for?

Iamlooker committed:
Move executeLiveData and killLiveData to CuberiteService

Iamlooker committed:
Migrate `updateLogsLivedata` to Kotlin Coroutine

Although this is not the best implementation but the lack of coroutines in the arch. might be reasoned for this.

Iamlooker committed:
Migrate `CuberiteService` to Kotlin Flows

Iamlooker commented:
Can you test this? I have migrated from livedata to flows. Once all is set I will move on to compose migration

mathiascode commented:
Seems to be working fine, and the changes look reasonable. Nice work!

mathiascode commented:
> Btw what is [this](https://github.com/Iamlooker/cuberite-android/blob/master/app/src/main/kotlin/org/cuberite/android/services/CuberiteService.kt#L157) tick for?

It's a hack for displaying an error when running an invalid executable. Could be replaced with something more appropriate at some point.

Iamlooker committed:
Migrate ControlScreen - 1

Iamlooker committed:
Migrate ControlScreen - 2

Iamlooker committed:
Fix wrong ipaddress

Iamlooker commented:
This is the current state of control screen:


https://github.com/cuberite/android/assets/60404093/e2a07311-5c83-4f20-bb9e-be57f7914660

Iamlooker committed:
Improve scaling animation

Iamlooker commented:
Would you also like to migrate to [Datastore](https://developer.android.com/topic/libraries/architecture/datastore) from SharedPreferences? This is also for [reference](https://android-developers.googleblog.com/2020/09/prefer-storing-data-with-jetpack.html).

I know this is too much going on in one PR but what are your thoughts?

mathiascode commented:
> This is the current state of control screen:
> 
>  sc.mp4

Could we keep UI changes in a separate PR after migration? I like the idea of displaying the IP address in the app itself too, but it would be easier to review and comment on details later.

Iamlooker commented:
Cool

Iamlooker commented:
I will remove these commits tomorrow

mathiascode commented:
> Would you also like to migrate to [Datastore](https://developer.android.com/topic/libraries/architecture/datastore) from SharedPreferences? This is also for [reference](https://android-developers.googleblog.com/2020/09/prefer-storing-data-with-jetpack.html).
> 
> I know this is too much going on in one PR but what are your thoughts?

No strong opinion. If it's the preferred replacement, and migrating existing preferences doesn't require too much work, then we could migrate later.

Iamlooker commented:
I will remove the ui commits tomorrow

Iamlooker commented:
I think the removing the commits will cause some issues with my local repository. Can we merge this in this state?

mathiascode commented:
Are you planning on migrating SettingsFragment in a separate PR? I'll try to review this PR next week, but I can't promise anything.

Iamlooker commented:
> Are you planning on migrating SettingsFragment in a separate PR? I'll try to review this PR next week, but I can't promise anything.

If you want I can do it in the same PR

mathiascode commented:
Sure, you could finish the migration in this PR if you have time.

ItsShadowCone commented:
> > I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.
> 
> Yes, but I did read about some changes in Android 13. I will look into this.

Apologies if I'm completely off the track (I haven't read up on this topic in years), but last I remembered there should still be an executable-mount tmpfs that should be accessible to apps iirc. We could use this to deploy the binary from data directory there immediately before launching it and immediately after cleaning it up (Linux keeps open files around even if you remove them from the file system)

Iamlooker committed:
Lay the foundations for SettingsPage

Iamlooker committed:
Improve Settings Screen

Only dialog boxes are remaining

Iamlooker committed:
Create Dialogs for Settings Screen

Iamlooker commented:
It should be ready to review now. If I missed anything please tell me. I left the old fragments in place, because navigation to Jetpack Compose Navigation is remaining. I will start doing that once the current changes are accepted

mathiascode commented:
Sorry for the delay. It will take some time to review, as I have other work taking up my time right now. I'm hoping to get this merged as soon as possible.

mathiascode commented:
> > > I should also mention that we're stuck on target API version 28, due to 29 no longer allowing executables in the data folder to run. It shouldn't be an issue in this case, but we will probably have to do something about it in a few years.
> > 
> > 
> > Yes, but I did read about some changes in Android 13. I will look into this.
> 
> Apologies if I'm completely off the track (I haven't read up on this topic in years), but last I remembered there should still be an executable-mount tmpfs that should be accessible to apps iirc. We could use this to deploy the binary from data directory there immediately before launching it and immediately after cleaning it up (Linux keeps open files around even if you remove them from the file system)

Do you have a link to documentation regarding this? It would be nice if there was a workaround, but I haven't seen this one mentioned before.

mathiascode commented on app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt:
```diff
@@ -0,0 +1,263 @@
+) {
+    val settings = rememberCategory(
+        title = stringResource(R.string.title_activity_settings),
+        icon = Icons.Rounded.Settings,
```
Webadmin icon is currently missing. Was this icon added to the wrong category?

mathiascode commented on app/src/main/kotlin/org/cuberite/android/ui/settings/SettingsScreen.kt:
```diff
@@ -0,0 +1,263 @@
+) {
+    val about = rememberCategory(
+        title = stringResource(R.string.settings_info_heading),
+        footerText = stringResource(R.string.settings_info_version, BuildConfig.VERSION_NAME),
```
The version label should open the F-Droid page when clicked.

Iamlooker replied:
Will add this in two days, I don't have pc right now 😔

mathiascode reviewed (commented):
Looks good in general. Seems to be working fine. I'll do some more testing on older Android versions and see if I spot any regressions.

Could you remove all unused resources, as well as the unused code in SettingsFragment? Would make it easier to see what the actual code diff is.

Iamlooker committed:
Remove unused settings code
Replace PreferenceFragment with Fragment

Iamlooker committed:
Remove unused resources

Iamlooker committed:
Add click listener to version footer

Iamlooker commented:
All aforementioned issues are fixed

Iamlooker commented:
Any update?

mathiascode commented:
Thanks for all the great work! Are you planning on removing fragments entirely?

Iamlooker commented:
Yes, I will create a PR to migrate to Compose Navigation 😉
